src/arch/arm/64/kernel/vspace.c

```diff
@@ -1621,8 +1663,29 @@ static exception_t decodeARMFrameInvocation(word_t invLabel, word_t length,
             current_syscall_error.type = seL4_IllegalOperation;
             return EXCEPTION_SYSCALL_ERROR;
         }
+#else
+        /* When in EL1, the mapping must be writeable for DC IVAC */
+        if (invLabel == ARMPageInvalidate_Data) {
+            lookupPTSlot_ret_t lu_ret = lookupPTSlot(find_ret.vspace_root, vaddr);
+            pte_t pte = *lu_ret.ptSlot;
+            void *base_ptr = (void *) cap_frame_cap_get_capFBasePtr(cap);
+            /* The cap mapping info could be out of date, so we need to check that the page we are
+               getting is the one the cap provides authority for. */
+            if (unlikely(lu_ret.ptBitsLeft != pageBitsForSize(cap_frame_cap_get_capFSize(cap)) ||
+                         !pte_is_page_type(pte) ||
+                         pte_get_page_base_address(pte) != pptr_to_paddr(base_ptr))) {
+                /* If the cap mapping info is out of date ignore the DC IVAC request silently to be
+                   consistent with ARMVSpaceInvalidate_Data. */
+                setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
+                return EXCEPTION_NONE;
```

Comment:
I would return an error saying it was done on a stale cap. If user space ever does this, it's probably unintended and a very hard to fix bug otherwise.

The situation is not the same as `ARMVSpaceInvalidate_Data`, because there we know the range is not mapped, but the range being unmapped is not an error because it is a vspace operation and it just means there's nothing to invalidate. While here the virtual address range may be mapped by another page cap, but doing cache operations on stale caps is wrong and `ARMPageInvalidate_Data` returns an error when called on unmapped page caps.

Reply:
You’re right, while the VSpace situation is also usually a sign of something not quite right, this is a more serious problem and it’s also not something anyone can have relied on yet, so an error is better. Will update.